### PR TITLE
[#IP-298] Add Subscription feed update logic based on servicePreferenceSetting mode

### PR DESCRIPTION
**Pull request by AleDore**
In order to complete the opt-in/opt-out service model implementation, this PR implements changes on profile's upsert workflow, with particular focus on impacts through Subscription Feed update activities, needed when considering new profile's model involving `servicePreferenceSettingsMode`.
#### List of Changes

- Update UpsertProfileOrchestrator
- Add unit tests and refactor existing ones

#### Motivation and Context

New service model requires changes to Profile Upsert operations in order to be compliant with the new model, by maintaining backward compatibility.

#### How Has This Been Tested?

Unit test

#### Types of changes

- [ ] Chore (nothing changes by a user perspective)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Checklist:

- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.

### CODE REVIEW

**Review by michaeldisaro — COMMENTED** [review]
LGTM apart minor, not mandatory, comment changes.

File: UpsertedProfileOrchestrator/__tests__/handler.test.ts [hunk 1]
```diff
@@ -251,4 +254,854 @@ describe("UpsertedProfileOrchestrator", () => {
...
+        getInput: jest.fn(() => upsertedProfileOrchestratorInput)
+      }
+    };
+
+    const orchestratorHandler = getUpsertedProfileOrchestratorHandler({
+      notifyOn: fromArray([expectedQueueName]).getOrElseL(undefined),
+      sendCashbackMessage: true
+    })(contextMockWithDf as any);
+
+    const result = orchestratorHandler.next();
```
Comment (gquadrati) [inline UpsertedProfileOrchestrator/__tests__/handler.test.ts:311]: For not calling `next` at each step you can either use `consumeGenerator` as visible here: https://github.com/pagopa/io-functions-pushnotifications/blob/a1506a9330b464b5989e8c119150bd401bca4d8c/utils/durable/utils.ts#L8

```suggestion
    const result = consumeGenerator(orchestratorHandler);
    expect(res).toBe(true);

    expect(contextMockWithDf.df.callActivityWithRetry).toBeCalledWith(
      "UpdateSubscriptionsFeedActivity",
      someRetryOptions,
      {
        fiscalCode: aFiscalCode,
        operation: "SUBSCRIBED",
        subscriptionKind: "PROFILE",
        updatedAt: upsertedProfileOrchestratorInput.updatedAt.getTime(),
        version: upsertedProfileOrchestratorInput.newProfile.version
      }
    );
```

**Review by BurnedMarshal — CHANGES_REQUESTED** [review]

File: UpsertedProfileOrchestrator/__tests__/handler.test.ts [hunk 2]
```diff
@@ -251,4 +254,854 @@ describe("UpsertedProfileOrchestrator", () => {
...
+  it("should call UpdateSubscriptionFeedActivity when profile is created", async () => {
+    const expectedQueueName = "queue_name" as NonEmptyString;
+    const upsertedProfileOrchestratorInput = UpsertedProfileOrchestratorInput.decode(
+      {
+        newProfile: {
+          ...aRetrievedProfile,
+          email: aEmailChanged, // Email changed to start the EmailValidationProcessOrchestrator
+          isInboxEnabled: true // Enable inbox to start the SendWelcomeMessagesActivity
+        },
+        oldProfile: undefined,
```
Comment (BurnedMarshal) [inline UpsertedProfileOrchestrator/__tests__/handler.test.ts:267]: This scenario should doesn't exists. When the profile is created `isInboxEnabled` is `false`. The next update with the  setting of `acceptedTosVersion` change it to `true`

**Review by gquadrati — APPROVED** [review]
LGTM except the comment for making tests more readable

File: UpsertedProfileOrchestrator/__tests__/handler.test.ts [hunk 3]
```diff
@@ -251,4 +254,854 @@ describe("UpsertedProfileOrchestrator", () => {
...
+      "EnqueueProfileCreationEventActivity",
+      someRetryOptions,
+      {
+        fiscalCode: aFiscalCode,
+        queueName: expectedQueueName
+      }
+    );
+  });
+
+  it("should call UpdateSubscriptionFeedActivity to unsubscribe the entire profile switching from MANUAL to AUTO", async () => {
```
Comment (BurnedMarshal) [inline UpsertedProfileOrchestrator/__tests__/handler.test.ts:712]:
```suggestion
  it("should call UpdateSubscriptionFeedActivity to unsubscribe the entire profile switching from AUTO to MANUAL", async () => {
```

File: UpsertedProfileOrchestrator/__tests__/handler.test.ts [hunk 4]
```diff
@@ -251,4 +254,854 @@ describe("UpsertedProfileOrchestrator", () => {
...
+      "EnqueueProfileCreationEventActivity",
+      someRetryOptions,
+      {
+        fiscalCode: aFiscalCode,
+        queueName: expectedQueueName
+      }
+    );
+  });
+
+  it("should call UpdateSubscriptionFeedActivity to subscribe the entire profile switching from AUTO to MANUAL", async () => {
```
Comment (BurnedMarshal) [inline UpsertedProfileOrchestrator/__tests__/handler.test.ts:833]:
```suggestion
  it("should call UpdateSubscriptionFeedActivity to subscribe the entire profile switching from MANUAL to AUTO", async () => {
```

File: UpsertedProfileOrchestrator/handler.ts [hunk 5]
```diff
@@ -202,48 +203,91 @@ export const getUpsertedProfileOrchestratorHandler = (params: {
...
+        }
+        // If Service Preference Mode doesn't change from oldProfile to newProfile in upsert operation
+        // or even if mode transition cause a no effect change (i.e from LEGACY to AUTO)
+        // we don't need to update Subscription Feed
+      } else if (
+        oldServicePreferenceMode !== newServicePreferencesMode &&
+        !(
+          oldServicePreferenceMode === ServicesPreferencesModeEnum.LEGACY &&
+          newServicePreferencesMode === ServicesPreferencesModeEnum.AUTO
+        )
```
[the remark this answers is not in the document]
Comment (michaeldisaro) [reply UpsertedProfileOrchestrator/handler.ts:267]: De Morgan says yes.

File: UpsertedProfileOrchestrator/handler.ts [hunk 6]
```diff
@@ -202,48 +203,91 @@ export const getUpsertedProfileOrchestratorHandler = (params: {
...
-            operation: "SUBSCRIBED",
-            serviceId: s as ServiceId,
-            subscriptionKind: "SERVICE",
-            updatedAt: updatedAt.getTime(),
-            version: newProfile.version
-          } as UpdateServiceSubscriptionFeedActivityInput
+      const { newServicePreferencesMode, oldServicePreferenceMode } = {
+        newServicePreferencesMode: newProfile.servicePreferencesSettings.mode.toString(),
+        oldServicePreferenceMode: oldProfile.servicePreferencesSettings.mode.toString()
+      };
```
Comment (gquadrati) [inline UpsertedProfileOrchestrator/handler.ts:209]:
```suggestion
        newServicePreferencesMode = newProfile.servicePreferencesSettings.mode.toString();
        oldServicePreferenceMode = oldProfile.servicePreferencesSettings.mode.toString();
```

**Review by BurnedMarshal — CHANGES_REQUESTED** [review]

File: UpsertedProfileOrchestrator/__tests__/handler.test.ts (hunk 2 above)
Comment (AleDore) [reply UpsertedProfileOrchestrator/__tests__/handler.test.ts:267]: Ok I'll remove it

File: UpsertedProfileOrchestrator/handler.ts (hunk 6 above)
Comment (AleDore) [reply UpsertedProfileOrchestrator/handler.ts:209]: Yes it's for avoid writing two instructions ;)

Comment (gquadrati) [reply UpsertedProfileOrchestrator/handler.ts:209]: I know :)

File: UpsertedProfileOrchestrator/__tests__/handler.test.ts [hunk 7]
```diff
@@ -241,7 +239,721 @@ describe("UpsertedProfileOrchestrator", () => {
...
+        },
+        oldProfile: {
+          ...aRetrievedProfile,
+          servicePreferencesSettings: autoProfileServicePreferencesSettings
+        },
+        updatedAt: new Date()
+      }
+    ).getOrElseL(_ =>
+      fail(
+        `Cannot decode UpsertedProfileOrchestratorInput: ${readableReport(_)}`
```
Comment (fabriziopapi) [inline UpsertedProfileOrchestrator/__tests__/handler.test.ts:270]: LTGM, but use _ for not used variable only

**Review by fabriziopapi — APPROVED** [review]

**Review by BurnedMarshal — APPROVED** [review]
LGTM

**Pull request merged**